presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/StateSerializer.java

```diff
@@ -0,0 +1,282 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.differentialentropy;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.function.AccumulatorStateSerializer;
+import com.facebook.presto.spi.type.Type;
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+
+import java.security.InvalidParameterException;
+
+import static com.facebook.presto.spi.StandardErrorCode.CONSTRAINT_VIOLATION;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.VarbinaryType.VARBINARY;
+import static java.lang.String.format;
+
+/**
+ * Serializes and creates differential-entropy states.
+ */
+public class StateSerializer
+        implements AccumulatorStateSerializer<State>
+{
+    public static StateStrategy create(
+            long size,
+            double sample)
+    {
+        return new UnweightedReservoirSampleStateStrategy(size);
+    }
+
+    public static StateStrategy create(
+            long size,
+            double sample,
+            double weight)
+    {
+        return new WeightedReservoirSampleStateStrategy(size);
+    }
+
+    public static StateStrategy create(
+            long size,
+            double sample,
+            double weight,
+            String method)
+    {
+        if (method.equalsIgnoreCase("reservoir_vasicek")) {
+            return new WeightedReservoirSampleStateStrategy(size);
+        }
+
+        throw new InvalidParameterException(format("unknown method %s or method/argument mismatch", method));
+    }
+
+    public static StateStrategy create(
+            long size,
+            double sample,
+            double weight,
+            String method,
+            double min,
+            double max)
+    {
+        if (method.equalsIgnoreCase("fixed_histogram_mle")) {
+            return new FixedHistogramMLEStateStrategy(size, min, max);
+        }
+
+        if (method.equalsIgnoreCase("fixed_histogram_jacknife")) {
+            return new FixedHistogramJacknifeStateStrategy(size, min, max);
+        }
+
+        throw new InvalidParameterException(format("unknown method %s or method/argument mismatch", method));
+    }
+
+    public static void validate(StateStrategy strategy, String method)
```

Comment:
This method is only used in DifferentialEntropyAggregation. Move it there and make it private.

Reply:
Added an explanation in the doc on the rationale for its being here. Will be happy to discuss further, of course.